### PR TITLE
Implement authorization API

### PR DESCRIPTION
**Pull request by Windawz**
The authorization microservice is currently being rewritten.

Several notes and guidelines were taken during a face-to-face review:
- Put repositories in data layer, make them return data objects
- Use the "Service" postfix for naming services
- Organize types by kind, then by purpose
- Hash passwords
- Request confirmation password during registration
- Name endpoints restfully
- Postfix WebAPI controller DTOs with "Request" and "Response"
- Adopt simpler naming
- Move to .NET 8.0 and a newer C# version
- Use the new `required` keyword to avoid assigning `null!` to properties
- Write less complicated code

More items may be added to this list in the future if necessary.

### CODE REVIEW

File: src/Microservices/Authorization/Api/Services/Implementations/LogInService.cs [hunk 1]
```diff
@@ -0,0 +1,39 @@
...
+        _tokenResponseMapperService = tokenResponseMapperService;
+    }
+
+    public async Task<TokenResponse> LogInAsync(LogInRequest request)
+    {
+        var (email, password) = _logInRequestMapperService.MapToEmailAndPassword(request);
+        var account = await _accountService.AccessAccountAsync(email, password);
+        var accessToken = await _accessTokenService.GenerateTokenAsync(account.Role);
+        var refreshToken = await _refreshTokenService.CreateRefreshTokenAsync(account.Role);
+        return _tokenResponseMapperService.MapFromTokenPair(accessToken, refreshToken);
```
Comment (MikhadziukAliaksei) [inline src/Microservices/Authorization/Api/Services/Implementations/LogInService.cs:37]: Make Mapper services => static

Comment (MikhadziukAliaksei) [reply src/Microservices/Authorization/Api/Services/Implementations/LogInService.cs:37]: For while solution

File: src/Microservices/Authorization/Api/Services/Mappers/Implementations/RefreshRequestMapperService.cs [hunk 2]
```diff
@@ -0,0 +1,21 @@
+using InnowiseClinic.Microservices.Authorization.Api.DataTransferObjects.Requests;
+using InnowiseClinic.Microservices.Authorization.Api.Services.Mappers.Interfaces;
+using InnowiseClinic.Microservices.Authorization.Application.Models;
+
+namespace InnowiseClinic.Microservices.Authorization.Api.Services.Mappers.Implementations;
+
+public class RefreshRequestMapperService : IRefreshRequestMapperService
+{
+    private readonly IRefreshTokenStringMapperService _refreshTokenStringMapperService;
```
Comment (MikhadziukAliaksei) [inline src/Microservices/Authorization/Api/Services/Mappers/Implementations/RefreshRequestMapperService.cs:9]: Merge ur mapper services and split by entities or layers

File: src/Microservices/Authorization/Application/Models/Role.cs [hunk 3]
```diff
@@ -0,0 +1,8 @@
+namespace InnowiseClinic.Microservices.Authorization.Application.Models;
+
+public enum Role
+{
+    Patient,
```
Comment (MikhadziukAliaksei) [inline src/Microservices/Authorization/Application/Models/Role.cs:5]: Number with 1 or the value 0 must be like None Unknown. Null value in this case will be mapped to Patient

File: src/Microservices/Authorization/Application/Services/Implementations/RefreshTokenService.cs [hunk 4]
```diff
@@ -0,0 +1,77 @@
...
+                await _refreshTokenRepository.SaveAsync();
+                replacementRefreshToken = await CreateRefreshTokenAsync(refreshToken.Role);
+            }
+        }
+
+        if (replacementRefreshToken is null)
+        {
+            throw new InvalidRefreshTokenException(refreshToken.TokenId);
+        }
+        else
```
Comment (MikhadziukAliaksei) [inline src/Microservices/Authorization/Application/Services/Implementations/RefreshTokenService.cs:67]: unused else typo for while solution

File: src/Microservices/Authorization/Application/Services/Interfaces/IRefreshTokenService.cs [hunk 5]
```diff
@@ -0,0 +1,12 @@
...
+using InnowiseClinic.Microservices.Authorization.Application.Services.Exceptions;
+
+namespace InnowiseClinic.Microservices.Authorization.Application.Services.Interfaces;
+
+public interface IRefreshTokenService
+{
+    Task<RefreshToken> CreateRefreshTokenAsync(Role role);
+    
+    /// <exception cref="InvalidRefreshTokenException"/>
+    Task<RefreshToken> CreateReplacementRefreshTokenAsync(RefreshToken refreshToken);
```
Comment (MikhadziukAliaksei) [inline src/Microservices/Authorization/Application/Services/Interfaces/IRefreshTokenService.cs:11]: weird name

File: src/Microservices/Authorization/Data/Repositories/Implementations/AccountRepository.cs [hunk 6]
```diff
@@ -0,0 +1,18 @@
...
+namespace InnowiseClinic.Microservices.Authorization.Data.Repositories.Implementations;
+
+public class AccountRepository : AsyncRepository<AccountEntity, AuthorizationDbContext>, IAccountRepository
+{
+    public AccountRepository(AuthorizationDbContext dbContext) : base(dbContext) { }
+
+    public async Task<AccountEntity?> GetAsync(string email)
+    {
+        return await DbContext.Accounts
+            .FirstOrDefaultAsync(account => EF.Functions.Like(account.Email, email));
```
Comment (MikhadziukAliaksei) [inline src/Microservices/Authorization/Data/Repositories/Implementations/AccountRepository.cs:16]: Why Like ef function?

File: src/Microservices/Shared/Api/ExceptionHandlers/ExceptionHandler.cs [hunk 7]
```diff
@@ -0,0 +1,61 @@
...
+    }
+
+    protected ILogger Logger { get; }
+    protected bool IsDevelopment { get; }
+
+    public async ValueTask<bool> TryHandleAsync(HttpContext httpContext, Exception exception, CancellationToken cancellationToken)
+    {
+        if (MapToStatusCode(exception) is int statusCode)
+        {
+            var problemDetails = _problemDetailsFactory.CreateProblemDetails(
```
Comment (MikhadziukAliaksei) [inline src/Microservices/Shared/Api/ExceptionHandlers/ExceptionHandler.cs:28]: problem details for bad request. Recreate GlobalExceptionHandler 
https://www.youtube.com/watch?v=H3EbflpXVmo

File: src/Microservices/Shared/Data/Repositories/Implementations/AsyncRepository`1.cs [hunk 8]
```diff
@@ -0,0 +1,50 @@
+using InnowiseClinic.Microservices.Shared.Data.Entities;
+using InnowiseClinic.Microservices.Shared.Data.Repositories.Interfaces;
+using Microsoft.EntityFrameworkCore;
+
+namespace InnowiseClinic.Microservices.Shared.Data.Repositories.Implementations;
+
+public abstract class AsyncRepository<TEntity, TContext> : IAsyncRepository<TEntity>
```
[the remark this answers is not in the document]
Comment (Windawz) [reply src/Microservices/Shared/Data/Repositories/Implementations/AsyncRepository`1.cs]: Fixed in 976670e, 12205e3.

File: src/Microservices/Shared/Data/Repositories/Implementations/AsyncRepository`1.cs [hunk 9]
```diff
@@ -0,0 +1,50 @@
...
+        
+        await Task.CompletedTask;
+    }
+
+    public async virtual Task DeleteAsync(TEntity entity)
+    {
+        DbContext.Set<TEntity>()
+            .Remove(entity);
+
+        await Task.CompletedTask;
```
[the remark this answers is not in the document]
Comment (Windawz) [reply src/Microservices/Shared/Data/Repositories/Implementations/AsyncRepository`1.cs]: Fixed by f71aa2c, 21bf64c.

File: src/Microservices/Shared/Data/Repositories/Implementations/AsyncRepository`1.cs [hunk 10]
```diff
@@ -0,0 +1,50 @@
...
+        await DbContext.Set<TEntity>()
+            .AddAsync(entity);
+    }
+
+    public async virtual Task UpdateAsync(TEntity entity)
+    {
+        DbContext.Set<TEntity>()
+            .Update(entity);
+        
+        await Task.CompletedTask;
```
[the remark this answers is not in the document]
Comment (Windawz) [reply src/Microservices/Shared/Data/Repositories/Implementations/AsyncRepository`1.cs]: Fixed by f71aa2c, 21bf64c.

File: src/Microservices/Shared/Data/Repositories/Interfaces/IAsyncRepository`1.cs [hunk 11]
```diff
@@ -0,0 +1,12 @@
+using InnowiseClinic.Microservices.Shared.Data.Entities;
+
+namespace InnowiseClinic.Microservices.Shared.Data.Repositories.Interfaces;
+
+public interface IAsyncRepository<TEntity> where TEntity : Entity
+{
+    Task<TEntity?> GetAsync(Guid id);
+    Task AddAsync(TEntity entity);
+    Task UpdateAsync(TEntity entity);
```
[the remark this answers is not in the document]
Comment (Windawz) [reply src/Microservices/Shared/Data/Repositories/Interfaces/IAsyncRepository`1.cs]: Fixed by f71aa2c, 21bf64c.

File: src/Microservices/Shared/Data/Repositories/Interfaces/IAsyncRepository`1.cs [hunk 12]
```diff
@@ -0,0 +1,12 @@
...
+    Task DeleteAsync(TEntity entity);
```
[the remark this answers is not in the document]
Comment (Windawz) [reply src/Microservices/Shared/Data/Repositories/Interfaces/IAsyncRepository`1.cs]: Fixed by f71aa2c, 21bf64c.

File: src/Microservices/Shared/Data/Entities/Configurations/EntityConfiguration.cs [hunk 13]
```diff
@@ -0,0 +1,15 @@
...
+namespace InnowiseClinic.Microservices.Shared.Data.Entities.Configurations;
+
+public class EntityConfiguration : IEntityTypeConfiguration<Entity>
+{
+    public void Configure(EntityTypeBuilder<Entity> builder)
+    {
+        builder.UseTpcMappingStrategy();
+        
+        builder.Property(entity => entity.Id)
+            .ValueGeneratedNever();
```
[the remark this answers is not in the document]
Comment (Windawz) [reply src/Microservices/Shared/Data/Entities/Configurations/EntityConfiguration.cs]: Fixed by 6ac8788, 51fb83f.

File: src/Microservices/Authorization/Api/Controllers/AccountsController.cs [hunk 14]
```diff
@@ -0,0 +1,54 @@
...
+
+namespace InnowiseClinic.Microservices.Authorization.Api.Controllers;
+
+[ApiController]
+[Route("accounts")]
+public class AccountsController : ControllerBase
+{
+    private readonly ILogInService _logInService;
+    private readonly IRegisterService _registerService;
+    private readonly IRegisterOtherService _registerOtherService;
```
[the remark this answers is not in the document]
Comment (Windawz) [reply src/Microservices/Authorization/Api/Controllers/AccountsController.cs]: Fixed by c61fc3d.

File: src/Microservices/Authorization/Api/Controllers/AccountsController.cs [hunk 15]
```diff
@@ -0,0 +1,54 @@
...
+        IRefreshService refreshService)
+    {
+        _logInService = logInService;
+        _registerService = registerService;
+        _registerOtherService = registerOtherService;
+        _refreshService = refreshService;
+    }
+
+    [HttpPost("login")]
+    public async Task<TokenResponse> LogIn(LogInRequest request)
```
[the remark this answers is not in the document]
Comment (Windawz) [reply src/Microservices/Authorization/Api/Controllers/AccountsController.cs]: Fixed by 7ba3bf7.

File: src/Microservices/Authorization/Api/Controllers/AccountsController.cs [hunk 16]
```diff
@@ -0,0 +1,54 @@
...
+    {
+        return await _logInService.LogInAsync(request);
+    }
+
+    [HttpPost("register")]
+    public async Task Register(RegisterRequest request)
+    {
+        await _registerService.RegisterAsync(request);
```
[the remark this answers is not in the document]
Comment (Windawz) [reply src/Microservices/Authorization/Api/Controllers/AccountsController.cs]: Fixed by 7ba3bf7.

File: src/Microservices/Authorization/Api/Program.cs [hunk 17]
```diff
@@ -0,0 +1,109 @@
...
+    public static void Main(string[] args)
+    {
+        var builder = WebApplication.CreateBuilder(args);
+
+        builder.Services.AddControllers();
+        builder.Services.AddEndpointsApiExplorer();
+        builder.Services.AddSwaggerGen();
+        builder.Services.AddAuthentication(JwtBearerDefaults.AuthenticationScheme)
+            .AddJwtBearer();
+        builder.Services.AddMappingExceptionHandler(options =>
```
[the remark this answers is not in the document]
Comment (Windawz) [reply src/Microservices/Authorization/Api/Program.cs]: Fixed by 222c717, aec5da9.

File: src/Microservices/Authorization/Api/Services/Implementations/LogInService.cs (hunk 1 above)
Comment (Windawz) [reply src/Microservices/Authorization/Api/Services/Implementations/LogInService.cs]: Fixed by 23c69c7.

File: src/Microservices/Authorization/Api/Services/Mappers/Implementations/RefreshRequestMapperService.cs (hunk 2 above)
Comment (Windawz) [reply src/Microservices/Authorization/Api/Services/Mappers/Implementations/RefreshRequestMapperService.cs]: Fixed by 23c69c7.

File: src/Microservices/Authorization/Application/Models/Role.cs (hunk 3 above)
Comment (Windawz) [reply src/Microservices/Authorization/Application/Models/Role.cs]: Fixed by c2a5736.

File: src/Microservices/Authorization/Application/Services/Interfaces/IRefreshTokenService.cs (hunk 5 above)
Comment (Windawz) [reply src/Microservices/Authorization/Application/Services/Interfaces/IRefreshTokenService.cs]: Fixed by 03adec1.

File: src/Microservices/Authorization/Data/Repositories/Implementations/AccountRepository.cs (hunk 6 above)
Comment (Windawz) [reply src/Microservices/Authorization/Data/Repositories/Implementations/AccountRepository.cs]: Fixed by 2d471b5, 40d70ab.

File: src/Microservices/Shared/Api/ExceptionHandlers/ExceptionHandler.cs (hunk 7 above)
Comment (Windawz) [reply src/Microservices/Shared/Api/ExceptionHandlers/ExceptionHandler.cs]: Possibly fixed by 222c717, 2c72696, fc1b6ad.

File: src/Microservices/Authorization/Application/Services/Implementations/RefreshTokenService.cs (hunk 4 above)
Comment (Windawz) [reply src/Microservices/Authorization/Application/Services/Implementations/RefreshTokenService.cs]: Fixed by 66fbe5d, 2ad4d46.

**Pull request merged**